src/basic/FStar.Errors.fs

```diff
@@ -833,31 +907,110 @@ let message_prefix =
+    (* The reporting of errors, particularly errors in the warn_error string itself
+       is delicate.
+       We keep a map from warn_error strings to their parsed results,
+         - Some list<error_setting> in case it parses and is interpreted successfully
+         - None in case it does not parse or is not intepretable
+    *)
+    let error_flags : BU.smap<(option<(list<error_setting>)>)> = BU.smap_create 10 in
+    (* set_error_flags is called by Options.set_options, parse_cmd_line etc,
+       upon parsing the options.
+       It parses the current warn_error string and sets the result in the
+       error_flags map above. In case it fails, it reports an Getopt error
+       for Options to report. Options may, in turn, report that error
+       back using the functionality of this module, e.g., log_issue *)
+    let set_error_flags () =
+        let parse (s:string) =
+          match !parser_callback with
+          | None -> failwith "Callback for parsing warn_error strings is not set"
+          | Some f -> f s
+        in
+        let we = Options.warn_error () in
+        try let r = parse we in
+            BU.smap_add error_flags we (Some r);
+            Getopt.Success
+        with Invalid_warn_error_setting msg ->
+            (BU.smap_add error_flags we None;
+             Getopt.Error ("Invalid --warn_error setting: " ^msg))
+    in
+    (* get_error_flags is called when logging an issue to figure out
+       which error level to report a particular issue at (Warning, Error etc.)
+       It is important that this function itself never raises an exception:
+       raising an error when trying to report an error is bad news, e.g., it
+       crashes the ide mode since it causes F* to exit abruptly.
+       So, we don't do any parsing here ... just look up the result of a
+       prior parse, falling back to the default settings in case the
+       parse didn't succeed *)
+    let get_error_flags () =
+      let we = Options.warn_error () in
+      match BU.smap_try_find error_flags we with
+      | Some (Some w) -> w
+      | _ -> default_settings
+    in
+    (* Setting the parser callback received from setup_hooks
+       and installing, in turn, callbacks in Options for
+       parsing warn_error settings and also for warning on the use of
+       unsafe options. *)
+    let set_callbacks (f:string -> list<error_setting>) =
+        parser_callback := Some f;
+        Options.set_error_flags_callback set_error_flags;
+        Options.set_option_warning_callback (warn_unsafe_options None)
+    in
+    set_callbacks, get_error_flags
+
+let lookup err =
+  let flags = error_flags () in
+  let with_tag tag =
+    let v, _, i = lookup_error flags err in
+    v, tag, i
+  in
+  match err with
+  | Warning_Defensive
+        when Options.defensive_fail() ->
+    with_tag CAlwaysError
+
+  | Warning_WarnOnUse
+        when Options.report_assumes () = Some "warn" ->
+    with_tag CWarning
```

Comment:
This bit means that doing `--warn_error @334` will be ignored if `--report_assumes warn` is on. So the following succeeds with just a warning:
```fstar
#set-options "--report_assumes warn"      
#set-options "--warn_error @334"                                 

let f (x:int) : False = admit()
```

Also, it seems I can downgrade `--report_assumes error` to `--report_assumes warn`. Maybe that shouldn't be the case? If I'm trying to check a big development for being admit-free, I would like to have set it as error throughout and not be cheated by them being turned to warnings (of which there might be plenty).

Reply:
thanks for the catch. i can revise